app/models/email.rb

```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class Email < ApplicationRecord
+  class Association < ApplicationRecord
+    belongs_to :email
+    belongs_to :object, polymorphic: true
+  end
+
+  has_many :associations, dependent: :destroy
+
+  UNDEFINED = Object.new
+  private_constant :UNDEFINED
+
+  scope :associated_with, ->(object) { where(id: Association.where(object: object).select(:email_id)) }
+  scope :tagged_with, ->(*tags) { tags.inject(self) { |scope, tag| scope.where("? = ANY (tags)", tag) } }
+
+  def create_association_with(*objects, as: nil) # rubocop:disable Naming/MethodParameterName
+    objects.each do |object|
+      Association.create!(
+        email: self,
+        object: object,
+        name: (as || object.model_name.singular),
+      )
+    end
+  end
```

Comment:
Doesn't look like we need `UNDEFINED` anymore.

Also, do we need `name` for anything on association? I don't see it used in any of the Email internals anymore?

Reply:
`name` is used within delviery_recorder: `create_association_with(*User.where(email: email.to).to_a, as: :recipient)`. It is to describer the relation between given object and the email.